cmd/perflock/perflock_test.go

```diff
@@ -0,0 +1,229 @@
+		if os.Getenv("GO_TEST_PROGRAM_MODE") != "" {
+			os.Setenv("GO_TEST_MODE", "program")
+		}
+		// Remove the flags registered by the testing package.
+		//
+		// NOTE: flag.Parse() has not been called yet (we're in TestMain), according
+		// to https://pkg.go.dev/testing
+		flag.CommandLine = flag.NewFlagSet("dummy", flag.ExitOnError) // This call is required because otherwise flags panics, if args are set between flag.Parse calls
+		main()
+
+	case "program": // Act like a "normal" program.
+		switch pmode := os.Getenv("GO_TEST_PROGRAM_MODE"); pmode {
+		case "sleeper":
+			sleeper()
+		default:
+			log.Fatalf("unknown program mode %q", pmode)
+		}
+	}
+}
+
+func waitForDaemon(t *testing.T, ctx context.Context, socket string) bool {
+	for {
+		c, err := net.Dial("unix", socket)
+		if err != nil {
+			// TODO(aktau): Deal with errors that aren't not found?
+			select {
+			case <-ctx.Done():
+				return false
+			case <-time.After(100 * time.Millisecond):
+				continue
+			}
+		}
+		c.Close()
+		return true
+	}
+}
+
+func TestExclusive(t *testing.T) {
+	t.Parallel()
+
+	socket := socketName(t)
+
+	// 1. Start a daemon.
+	mustStartDaemon(t, socket)
+
+	// 2. Start three sleepers in EXCLUSIVE mode, each sleeping for 0.5s.
+	start := time.Now()
+	var sleepers [3]*exec.Cmd
+	for i := range sleepers {
+		sleepers[i] = mustStartSleeper(t, socket)
+	}
+
+	// 3. Wait for them all to finish.
+	for _, sleeper := range sleepers {
+		sleeper.Wait()
+	}
+
+	// Assert that they ran sequentially by making sure it took longer than
+	// sleep_time*num_sleepers.
+	if got, want := time.Since(start), time.Duration(len(sleepers))*sleepDuration; got < want {
+		t.Errorf("expected %d sleepers each sleeping %v to run sequentially, but time passed is %v",
+			len(sleepers), sleepDuration, got)
+	}
+}
+
+func TestShared(t *testing.T) {
+	t.Parallel()
+
+	socket := socketName(t)
+
+	// 1. Start a daemon.
+	mustStartDaemon(t, socket)
+
+	// 2. Start three sleepers in SHARED mode, each sleeping for 0.5s.
+	start := time.Now()
+	var sleepers [3]*exec.Cmd
+	for i := range sleepers {
+		sleepers[i] = mustStartSleeper(t, socket, "-shared")
+	}
+
+	for _, sleeper := range sleepers {
+		sleeper.Wait()
+	}
+
+	// Assert that they ran concurrently by making sure it was shorter than than
+	// sleep_time*num_sleepers.
+	if got, maxTime := time.Since(start), time.Duration(len(sleepers))*sleepDuration; got > maxTime {
+		t.Errorf("expected %d shared sleepers each sleeping %v to not take as long as them sleeping sequentially, but time passed is %v",
+			len(sleepers), sleepDuration, got)
+	}
+}
+
+// funcname returns the function name of the caller.
+func funcname(skip int) string {
+	var pcs [1]uintptr
+	if runtime.Callers(skip+1, pcs[:]) != 1 {
+		return "UNKNOWN"
+	}
+	fr, _ := runtime.CallersFrames(pcs[:]).Next()
+	return fr.Func.Name()
+}
+
+// socketName returns a unique socket name per test.
+func socketName(t *testing.T) string {
+	if runtime.GOOS == "linux" && false {
```

Comment:
I see the `&& false`. Should this path be deleted? Or revived?

Reply:
Sorry, that was a leftover from when I was testing that both paths still work on Linux (meaning I could be more sure that it would work on Darwin). Removed it now.